File: GanttView/RadGanttViewTruckSchedulerExample/RadGanttViewExample/CustomGanttGraphicalViewBaseTaskElement.cs
using System;
using System.Data;
using System.Drawing;
using Telerik.WinControls;
using Telerik.WinControls.Paint;
using Telerik.WinControls.UI;

namespace RadGanttViewExample
{
    public class CustomGanttGraphicalViewBaseTaskElement : GanttGraphicalViewBaseTaskElement
    {
        private Random rnd = new Random();
        private DrivingToPickUpLocationElement drivingToPickUpLocationElement;
        private LoadingElement loadingElement;
        private DrivingElement drivingElement;
        private DriverRestElement driverRestElement;
        private WaitingElement waitingElement;
        private UnloadingElement unloadingElement;
        private LightVisualElement borderElement;

        protected override void InitializeFields()
        {
            base.InitializeFields();

            this.TextAlignment = ContentAlignment.MiddleLeft;
            this.DrawFill = true;
            this.GradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.BackColor = Color.LightGray;
            this.DrawBorder = true;
            this.BorderGradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.BorderColor = Color.Black;
        }

        protected override void CreateChildElements()
        {
            base.CreateChildElements();

            this.drivingToPickUpLocationElement = new DrivingToPickUpLocationElement();
            this.drivingToPickUpLocationElement.DrawFill = true;
            this.drivingToPickUpLocationElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.drivingToPickUpLocationElement.BackColor = Color.LightBlue;

            this.loadingElement = new LoadingElement();
            this.loadingElement.DrawFill = true;
            this.loadingElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.loadingElement.BackColor = Color.DarkBlue;

            this.drivingElement = new DrivingElement();
            this.drivingElement.DrawFill = true;
            this.drivingElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.drivingElement.BackColor = Color.Green;

            this.driverRestElement = new DriverRestElement();
            this.driverRestElement.DrawFill = true;
            this.driverRestElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.driverRestElement.BackColor = Color.SlateGray;

            this.waitingElement = new WaitingElement();
            this.waitingElement.DrawFill = true;
            this.waitingElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.waitingElement.BackColor = Color.Gold;

            this.unloadingElement = new UnloadingElement();
            this.unloadingElement.DrawFill = true;
            this.unloadingElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.unloadingElement.BackColor = Color.RosyBrown;

            this.borderElement = new LightVisualElement();
            this.borderElement.DrawBorder = true;
            this.borderElement.BorderGradientStyle = Telerik.WinControls.GradientStyles.Solid;
            this.borderElement.BorderColor = Color.Black;
            this.borderElement.ShouldHandleMouseInput = false;
            this.borderElement.NotifyParentOnMouseInput = true;

            this.Children.Add(this.drivingToPickUpLocationElement);
            this.Children.Add(this.loadingElement);
            this.Children.Add(this.drivingElement);
            this.Children.Add(this.driverRestElement);
            this.Children.Add(this.waitingElement);
            this.Children.Add(this.unloadingElement);
            this.Children.Add(this.borderElement);
        }

        protected override void PaintElement(Telerik.WinControls.Paint.IGraphics graphics, float angle, SizeF scale)
        {
            base.PaintElement(graphics, angle, scale);

            PointF[] points = new PointF[4];
            points[0] = new PointF(this.loadingElement.BoundingRectangle.X, this.Size.Height - 1);
            points[1] = new PointF(this.loadingElement.BoundingRectangle.Right, this.Size.Height * 0.35f);
            points[2] = new PointF(this.unloadingElement.BoundingRectangle.X, this.Size.Height * 0.35f);
            points[3] = new PointF(this.unloadingElement.BoundingRectangle.Right, this.Size.Height - 1);

            graphics.ChangeSmoothingMode(System.Drawing.Drawing2D.SmoothingMode.AntiAlias);
            graphics.FillPolygon(Color.FromArgb(100, Color.Gray), points);
            graphics.RestoreSmoothingMode();
        }
        
        protected override SizeF ArrangeOverride(SizeF finalSize)
        {
            CustomGanttViewTaskItemElement itemElement = (CustomGanttViewTaskItemElement)this.Parent;
            DataRowView boundItem = itemElement.Data.DataBoundItem as DataRowView;
            GanttViewGraphicalViewElement graphicalView = itemElement.GraphicalViewElement;

            SizeF availableSize = base.ArrangeOverride(finalSize);
            RectangleF clientRect = new RectangleF(PointF.Empty, availableSize);
            RectangleF arrangeRect = new RectangleF(clientRect.X, clientRect.Y, 0, clientRect.Height * 0.3f);

            float arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add((TimeSpan)boundItem["DrivingToPickUpLocation"])).Width;
            arrangeRect.Width = arrangeRectWidth;
            this.drivingToPickUpLocationElement.Arrange(arrangeRect);

            arrangeRect.X += arrangeRectWidth;
            arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add((TimeSpan)boundItem["Loading"])).Width;
            arrangeRect.Width = arrangeRectWidth;
            this.loadingElement.Arrange(arrangeRect);

            arrangeRect.X += arrangeRectWidth;
            arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add((TimeSpan)boundItem["Driving"])).Width;
            arrangeRect.Width = arrangeRectWidth;
            this.drivingElement.Arrange(arrangeRect);

            arrangeRect.X += arrangeRectWidth;
            arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add((TimeSpan)boundItem["DriverRest"])).Width;
            arrangeRect.Width = arrangeRectWidth;
            this.driverRestElement.Arrange(arrangeRect);

            arrangeRect.X += arrangeRectWidth;
            arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add((TimeSpan)boundItem["Waiting"])).Width;
            arrangeRect.Width = arrangeRectWidth;
            this.waitingElement.Arrange(arrangeRect);

            arrangeRect.X += arrangeRectWidth;
            arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add((TimeSpan)boundItem["Unloading"])).Width;
            arrangeRect.Width = arrangeRectWidth;
            this.unloadingElement.Arrange(arrangeRect);

            arrangeRect.X = clientRect.X;
            arrangeRect.Width = clientRect.Width;
            this.borderElement.Arrange(arrangeRect);

            return availableSize;
        }
    }

    public class DrivingToPickUpLocationElement : LightVisualElement
    { }

    public class LoadingElement : LightVisualElement
    { }

    public class DrivingElement : LightVisualElement
    { }

    public class DriverRestElement : LightVisualElement
    { }

    public class WaitingElement : LightVisualElement
    { }

    public class UnloadingElement : LightVisualElement
    { }
}

File: GanttView/RadGanttViewTruckSchedulerExample/RadGanttViewExampleVB/CustomGanttGraphicalViewBaseTaskElement.vb
Imports Telerik.WinControls.UI

Public Class CustomGanttGraphicalViewBaseTaskElement
    Inherits GanttGraphicalViewBaseTaskElement

    Private rnd As Random = New Random()
    Private drivingToPickUpLocationElement As DrivingToPickUpLocationElement
    Private loadingElement As LoadingElement
    Private drivingElement As DrivingElement
    Private driverRestElement As DriverRestElement
    Private waitingElement As WaitingElement
    Private unloadingElement As UnloadingElement
    Private borderElement As LightVisualElement

    Protected Overrides Sub InitializeFields()
        MyBase.InitializeFields()
        Me.TextAlignment = ContentAlignment.MiddleLeft
        Me.DrawFill = True
        Me.GradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.BackColor = Color.LightGray
        Me.DrawBorder = True
        Me.BorderGradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.BorderColor = Color.Black
    End Sub

    Protected Overrides Sub CreateChildElements()
        MyBase.CreateChildElements()
        Me.drivingToPickUpLocationElement = New DrivingToPickUpLocationElement()
        Me.drivingToPickUpLocationElement.DrawFill = True
        Me.drivingToPickUpLocationElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.drivingToPickUpLocationElement.BackColor = Color.LightBlue
        Me.loadingElement = New LoadingElement()
        Me.loadingElement.DrawFill = True
        Me.loadingElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.loadingElement.BackColor = Color.DarkBlue
        Me.drivingElement = New DrivingElement()
        Me.drivingElement.DrawFill = True
        Me.drivingElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.drivingElement.BackColor = Color.Green
        Me.driverRestElement = New DriverRestElement()
        Me.driverRestElement.DrawFill = True
        Me.driverRestElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.driverRestElement.BackColor = Color.SlateGray
        Me.waitingElement = New WaitingElement()
        Me.waitingElement.DrawFill = True
        Me.waitingElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.waitingElement.BackColor = Color.Gold
        Me.unloadingElement = New UnloadingElement()
        Me.unloadingElement.DrawFill = True
        Me.unloadingElement.GradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.unloadingElement.BackColor = Color.RosyBrown
        Me.borderElement = New LightVisualElement()
        Me.borderElement.DrawBorder = True
        Me.borderElement.BorderGradientStyle = Telerik.WinControls.GradientStyles.Solid
        Me.borderElement.BorderColor = Color.Black
        Me.borderElement.ShouldHandleMouseInput = False
        Me.borderElement.NotifyParentOnMouseInput = True
        Me.Children.Add(Me.drivingToPickUpLocationElement)
        Me.Children.Add(Me.loadingElement)
        Me.Children.Add(Me.drivingElement)
        Me.Children.Add(Me.driverRestElement)
        Me.Children.Add(Me.waitingElement)
        Me.Children.Add(Me.unloadingElement)
        Me.Children.Add(Me.borderElement)
    End Sub

    Protected Overrides Sub PaintElement(ByVal graphics As Telerik.WinControls.Paint.IGraphics, ByVal angle As Single, ByVal scale As SizeF)
        MyBase.PaintElement(graphics, angle, scale)
        Dim points As PointF() = New PointF(3) {}
        points(0) = New PointF(Me.loadingElement.BoundingRectangle.X, Me.Size.Height - 1)
        points(1) = New PointF(Me.loadingElement.BoundingRectangle.Right, Me.Size.Height * 0.35F)
        points(2) = New PointF(Me.unloadingElement.BoundingRectangle.X, Me.Size.Height * 0.35F)
        points(3) = New PointF(Me.unloadingElement.BoundingRectangle.Right, Me.Size.Height - 1)
        graphics.ChangeSmoothingMode(System.Drawing.Drawing2D.SmoothingMode.AntiAlias)
        graphics.FillPolygon(Color.FromArgb(100, Color.Gray), points)
        graphics.RestoreSmoothingMode()
    End Sub

    Protected Overrides Function ArrangeOverride(ByVal finalSize As SizeF) As SizeF
        Dim itemElement As CustomGanttViewTaskItemElement = CType(Me.Parent, CustomGanttViewTaskItemElement)
        Dim boundItem As DataRowView = TryCast(itemElement.Data.DataBoundItem, DataRowView)
        Dim graphicalView As GanttViewGraphicalViewElement = itemElement.GraphicalViewElement
        Dim availableSize As SizeF = MyBase.ArrangeOverride(finalSize)
        Dim clientRect As RectangleF = New RectangleF(PointF.Empty, availableSize)
        Dim arrangeRect As RectangleF = New RectangleF(clientRect.X, clientRect.Y, 0, clientRect.Height * 0.3F)
        Dim arrangeRectWidth As Single = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add(CType(boundItem("DrivingToPickUpLocation"), TimeSpan))).Width
        arrangeRect.Width = arrangeRectWidth
        Me.drivingToPickUpLocationElement.Arrange(arrangeRect)
        arrangeRect.X += arrangeRectWidth
        arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add(CType(boundItem("Loading"), TimeSpan))).Width
        arrangeRect.Width = arrangeRectWidth
        Me.loadingElement.Arrange(arrangeRect)
        arrangeRect.X += arrangeRectWidth
        arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add(CType(boundItem("Driving"), TimeSpan))).Width
        arrangeRect.Width = arrangeRectWidth
        Me.drivingElement.Arrange(arrangeRect)
        arrangeRect.X += arrangeRectWidth
        arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add(CType(boundItem("DriverRest"), TimeSpan))).Width
        arrangeRect.Width = arrangeRectWidth
        Me.driverRestElement.Arrange(arrangeRect)
        arrangeRect.X += arrangeRectWidth
        arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add(CType(boundItem("Waiting"), TimeSpan))).Width
        arrangeRect.Width = arrangeRectWidth
        Me.waitingElement.Arrange(arrangeRect)
        arrangeRect.X += arrangeRectWidth
        arrangeRectWidth = graphicalView.GetDrawRectangle(itemElement.Data, itemElement.Data.Start, itemElement.Data.Start.Add(CType(boundItem("Unloading"), TimeSpan))).Width
        arrangeRect.Width = arrangeRectWidth
        Me.unloadingElement.Arrange(arrangeRect)
        arrangeRect.X = clientRect.X
        arrangeRect.Width = clientRect.Width
        Me.borderElement.Arrange(arrangeRect)
        Return availableSize
    End Function
End Class

Public Class DrivingToPickUpLocationElement
    Inherits LightVisualElement
End Class

Public Class LoadingElement
    Inherits LightVisualElement
End Class

Public Class DrivingElement
    Inherits LightVisualElement
End Class

Public Class DriverRestElement
    Inherits LightVisualElement
End Class

Public Class WaitingElement
    Inherits LightVisualElement
End Class

Public Class UnloadingElement
    Inherits LightVisualElement
End Class